packages/financial_analysis/cli.py

```diff
@@ -437,19 +445,119 @@ def cmd_review_transaction_categories(
+                if not positions:
+                    continue
+                # Build identifiers for DB lookup
+                group_eids: list[str] = []
+                group_fps: list[str] = []
+                group_items: list[_ReviewPreparedItem] = []
+                for i in positions:
+                    tx = ctv_items[i]
+                    eid = str(tx.get("id")).strip() if tx.get("id") is not None else None
+                    if eid:
+                        group_eids.append(eid)
+                    fp = compute_fingerprint(source_provider=source_provider, tx=tx)
+                    group_fps.append(fp)
+                    group_items.append(
+                        _ReviewPreparedItem(
+                            pos=i,
+                            tx=tx,
+                            suggested="",  # not used in persistence path
+                            external_id=eid,
+                            fingerprint=fp,
+                        )
+                    )
+
+                # Query duplicates once per group; when unanimous, auto-apply
+                _dupes, unanimous = _review_query_group_duplicates(
+                    session,
+                    source_provider=source_provider,
+                    source_account=source_account,
+                    group_eids=group_eids,
+                    group_fps=group_fps,
+                    exemplars=1,
+                )
+                if not unanimous:
+                    continue
+
+                _review_persist_group(
+                    session,
+                    source_provider=source_provider,
+                    source_account=source_account,
+                    group_items=group_items,
+                    final_cat=unanimous,
+                    category_source="rule",
+                )
+                session.commit()
+
+                for i in positions:
+                    prefilled_positions.add(i)
+                prefilled_groups += 1
+
+    except Exception as e:
+        print(f"Error: DB prefill failed: {e}", file=sys.stderr)
+        return 1
+
+    # Build unresolved subset in original order
+    unresolved_indices: list[int] = [i for i in range(total) if i not in (prefilled_positions)]
+    unresolved_ctv: list = [ctv_items[i] for i in unresolved_indices]
+
+    # When everything was resolved from DB, jump straight to the (no-op) review
+    # with an empty list so the pre-review summary/exit path remains familiar.
+    if len(unresolved_ctv) == 0:
+        try:
+            review_transaction_categories(
+                [],
+                source_provider=source_provider,
+                source_account=source_account,
+                database_url=database_url,
+                prefilled_groups=prefilled_groups,
+                allow_create=allow_create,
+            )
+        except Exception as e:
+            print(f"Error: review failed: {e}", file=sys.stderr)
+            return 1
```

Comment:
@CharlieHelps this is A LOT of code to inline. move this to its own helper function and make sure it's descriptively named.